Migrate Customer view page

sarjon opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migrates Customers information page.
| Type?         | new feature
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/11010
| How to test?  | Access `admin-dev/index.php/sell/customers/1/view` and compare page with legacy one. **NOTE: it does not implement _"Transform to a customer account"_ and _"Add a private note"_ features to keep this PR manageable. It will be done in separate PR.** 

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/10992)
<!-- Reviewable:end -->

sarjon committed:
init customers view

sarjon committed:
define contracts for customer information query

sarjon committed:
add personal customer information collecting

sarjon committed:
add customer order information

sarjon commented:
@matks can you take a look at this wip pr? i dont know if its okay to use that many dto object, wdyt?

sarjon committed:
add customer carts information

matks commented:
@sarjon Yes, this page is complex so using DTOs here is a good idea. The alternative would be to have either an insane number of function arguments or big arrays that provide no guarantee about their content.

However we can try to make these DTOs very easy-to-use and easy-to-extend for module developers. One idea I can think about is to add a `createFromArray()` static function for them.


This allows a module developers that is used to manage customer order data by arrays to do this:
```
$this_is_my_array_I_love_to_use = array(
    'orderId' => ...,
    'orderPlacedDate' => ...,
    'paymentMethodName'=>...
);

// now the migrated page in SF wants a DTO ? okay I'll do that:
$this_is_the_dto_I_have_to_build = CustomerOrderInformation::createFromArray($this_is_my_array_I_love_to_use);

// yay ! this new concept "DTO" is not painful to handle
```

sarjon committed:
add customer products data collecting

sarjon committed:
collect customer messages information

sarjon committed:
collect customer discounts data

sarjon committed:
collect information about last emails sent to customer

sarjon committed:
collect last connections of customer

sarjon committed:
add customer group information

sarjon committed:
add customer referrers information collecting

sarjon committed:
cs fixer

sarjon committed:
add customer adddress information collecting

sarjon committed:
remove redundant suffixes

sarjon committed:
cs fixer

sarjon committed:
start building view ui

sarjon committed:
wip customer information display

tomas862 commented on src/Core/Domain/Customer/Exception/CustomerException.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Class CustomerException is base "Customer" context exception.
+ */
+class CustomerException extends \Exception
```
Why not extending DomainException?

```suggestion
class CustomerException extends DomainException
```

sarjon committed:
finish building personal information view block

sarjon committed:
add order information block

sarjon committed:
add carts information block

sarjon committed:
add bought products information block

sarjon committed:
add viewed products information

sarjon committed:
add customer addresses information block

sarjon committed:
add private note block

sarjon committed:
add messages information block

sarjon committed:
add discounts information block

sarjon committed:
add sent emails information block

sarjon committed:
add last connections information

sarjon committed:
add groups information display

sarjon committed:
update templates naming

sarjon committed:
minor fixes to customers view layout

sarjon committed:
fix listing of discounts

tomas862 committed:
fix CustomerException parent

Co-Authored-By: sarjon <jonusas.sarunas@gmail.com>

sarjon committed:
fix namespace for exception

sarjon committed:
minor ui changes

sarjon committed:
fix translations

sarjon committed:
add security and demo checks

sarjon committed:
handle customer not found exception

sarjon committed:
cs fixer

sarjon committed:
fix order date format

matks commented on src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php:
```diff
@@ -0,0 +1,590 @@
+
+        $this->assertCustomerWasFound($customerId, $customer);
+
+        Context::getContext()->customer = $customer;
```
Woah 😮 legacy behavior ?

sarjon replied:
yeah, you must set customer to `Context` in order to make it work. :/

matks commented on src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php:
```diff
@@ -0,0 +1,590 @@
+        $customerId = $query->getCustomerId();
+        $customer = new Customer($customerId->getValue());
+
+        $this->assertCustomerWasFound($customerId, $customer);
```
👍

matks commented on src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php:
```diff
@@ -0,0 +1,590 @@
+
+        if ($totalPaid = Db::getInstance()->getValue($sql)) {
+            $sql = '
+                SELECT SQL_CALC_FOUND_ROWS COUNT(*)
```
No mistake ? `SQL_CALC_FOUND_ROWS` along with `COUNT(*)` ?

sarjon replied:
well... i thought about trying to refactor it, but dont know if its worth, do you? it will be refactored later at some point and this code works for now.

matks replied:
We need to improve this, following [the boy scout rule](https://deviq.com/boy-scout-rule/) 😉

matks commented on src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php:
```diff
@@ -0,0 +1,590 @@
+
+            Db::getInstance()->getValue($sql);
+
+            return (int) Db::getInstance()->getValue('SELECT FOUND_ROWS()') + 1;
```
And you use `FOUND_ROWS` after `SQL_CALC_FOUND_ROWS` and `COUNT(*)` 🤔 . I'm sorry I don't get it.
Even if this comes from legacy code, we need to improve it.

matks replied:
🆙

matks commented on src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php:
```diff
@@ -0,0 +1,590 @@
+
+            if (!isset($order['order_state'])) {
+                $order['order_state'] = $this->translator->trans(
+                    'There is no status defined for this order.',
```
Should this be here or in the Twig template ? Is it really an `order_state` property of the Order, or is it a display rule only for templating ?

sarjon replied:
i think its for templating, but since query `GetCustomerInformation` is all about getting view data, maybe its ok? :thinking: if you think no, ill update. :)

matks replied:
I think this will be better in the template

matks commented on src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php:
```diff
@@ -0,0 +1,590 @@
+
+            if ($order['valid']) {
+                $validOrders[] = $customerOrderInformation;
+                $totalSpent += $order['total_paid_real_not_formated'] / $order['conversion_rate'];
```
Do not assume `$order['conversion_rate']` will be an integer different from zero. Assert it 😉

sarjon replied:
it must be. :) it comes directly from db and it's `decimal(13,6)`.

matks replied:
Still cant assume it. Do not rely on your database structure. Trust no one 😛

sarjon replied:
well, it was working for years (probably :D), so i think it safe to leave it like it is, at some point it will be totally refactored along with other adapters. :)

however, if you really want me to assert it, then what fallback action should it take if it fails?

matks replied:
We can assume that if the data is not valid, although it comes from the database, it must be
- either a developer modifying the code, so it's a mistake
- or a database structure not valid
In both cases we should throw an Exception as these are critical scenarii so we can fail the request. [Crash early](https://www.quora.com/What-does-crash-early-mean-in-software-engineering).

sarjon replied:
ok, but what exception should be thrown here? its part of domain, but i dont think its neither `CustomerException` nor `DomainException` would be correct here, maybe global `LogicException`, wdyt?

matks commented on src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php:
```diff
@@ -0,0 +1,590 @@
+
+        foreach ($carts as $cart) {
+            $cart = new Cart((int) $cart['id_cart']);
+            Context::getContext()->cart = $cart;
```
What is the purpose of setting `Context::getContext()->cart` in this `foreach` loop ?

sarjon replied:
to make legacy things work. :D i know it sucks, but in this case it must be set to make  `$cart->getSummaryDetails();` work. :/

matks commented on src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php:
```diff
@@ -0,0 +1,590 @@
+            );
+
+            if (!Validate::isLoadedObject($product)) {
+                continue;
```
Should we `continue` ... or display a warning message (although I dont know how - exception ? flash message ?) ? It might be a PM decision.

sarjon replied:
i think not a flash message for sure, because command can be dispatched using console application as well.

matks replied:
Since this is a view, I would not want it to crash because of a bad product.
But we can log the issue at least :) so that it's easier to debug if someone is wondering "why my product X is not displayed in my customer view ?"

sarjon replied:
hmm, i think it may be done this way on purpose.

if admin deletes a product that was previously viewed by customers, i dont think he cares any more about it, right? so why would you want to have history of all products (even those who were deleted year or more ago) that customer viewed? wdyt?

matks commented on src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php:
```diff
@@ -0,0 +1,590 @@
+
+            $customerMessages[] = new MessageInformation(
+                (int) $message['id_customer_thread'],
+                substr(strip_tags(html_entity_decode($message['message'], ENT_NOQUOTES, 'UTF-8')), 0, 75),
```
😅

sarjon replied:
i think this can be done in twig? :D

matks replied:
Is this processing intended to *decode* the stored data or is it intended to *display properly* the data in an HTML page ?
If answer is 1, we keep it here. If answer is 2, this goes in the template

sarjon replied:
to be honest, i dont really know why it was done this way. :/ i've simplified it a little.

matks commented on src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php:
```diff
@@ -0,0 +1,590 @@
+        $customerMessages = [];
+        $messages = CustomerThread::getCustomerMessages((int) $customer->id);
+
+        $messageStatues = [
```
`$messageStatues` => `$messageStatuses`

matks reviewed (changes requested):
1st review (there will be more): reviewed `GetCustomerInformationHandler`

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Customer/Blocks/View/personal_information.html.twig:
```diff
@@ -0,0 +1,184 @@
+        </div>
+        <div class="col-8">
+          {% if customerInformation.generalInformation.customerBySameEmailExists %}
+            <p>{{ 'A registered customer account using the defined email address already exists. '|trans({}, 'Admin.Orderscustomers.Notification') }}</p>
```
Watch the space at the end of the wording, unless it is made on purpose?

sarjon replied:
translation key contains this space. :/

matks commented on src/Core/Domain/Customer/Dto/CartInformation.php:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * @var string
+     */
+    private $cartId;
```
String, not int ?

sarjon replied:
nope, its formatted car id like "000001" for id 1. do you think it can be moved to twig?

matks commented on src/Core/Domain/Customer/Dto/CustomerInformation.php:
```diff
@@ -0,0 +1,249 @@
+    /**
+     * @var OrdersInformation
+     */
+    private $ordersInformation;
```
When having long sections of private vars statements, you can use "list" style:
```
/** @var CustomerId */
private $customerId;
/** @var PersonalInformation */
private $personalInformation;
/** @var OrdersInformation */
private $ordersInformation;
...
```

matks commented on src/Core/Domain/Customer/Dto/OrdersInformation.php:
```diff
@@ -0,0 +1,84 @@
+     * @param OrderInformation[] $validOrders
+     * @param OrderInformation[] $invalidOrders
+     */
+    public function __construct($totalSpent, array $validOrders, array $invalidOrders)
```
Can we add smoke tests for DTOs ? It would be a test that creates the object, and calls 2 getters "just to check it works". It ensures that, if someone modifies the class and introduces a syntax error (missing `;`), it gets caught immediately. It's a very basic test that allows to perform big IDE-driven refactoring such as namespace renaming or class/methods renaming 😃 

Exemple
```
class OrdersInformationTest extends TestCase
{
    public function testCanBeBuilt()
    {
        $a = new OrdersInformation(10, [], []);
        $this->assertEquals('a', $a-> getTotalSpent());
    }
}
```

matks replied:
Same for all DTO

sarjon replied:
yes, sure. :+1:

matks replied:
I added a checkbox here https://github.com/PrestaShop/PrestaShop/issues/11010#issue-370126654
so once we have merged this PR we can do a 2nd PR to do these tests

matks commented on src/Core/Domain/Customer/Dto/OrdersInformation.php:
```diff
@@ -0,0 +1,84 @@
+    private $invalidOrders;
+
+    /**
+     * @var string
```
string, not int ?

sarjon replied:
its a string as it is formatted price like "$25.36".

matks reviewed (changes requested):
2nd review done :)

sarjon commented:
@matks currently command is called `GetCustomerInformation` but now i think that `GetCustomerForViewing` name would be more appropriate as it reflects "View" action in Back Office, wdyt?

sarjon committed:
fix typo

matks commented:
@sarjon You are right, this data is definitely "enriched" for view so this name is better 👍

sarjon commented:
@matks i've addressed all your comments, there are still few minor questions to resolve but i dont think this will change PR a lot, so maybe you can add `waiting for QA`? :)

matks commented:
@sarjon Yes 😉 we only got a small `lint` step for Travis to go green

marionf commented:
Hello @sarjon 

In "already see" block, product's name are not correctly displayed when there is accents

![capture d ecran_623](https://user-images.githubusercontent.com/13449658/48542909-9ec6e580-e8c0-11e8-91ec-52a8a3ba96cb.png)

In "orders" block, when I click to display an order, I am redirected to the list of orders instead of the specific order. In the link I see "viewcustomer", before it was "vieworder".

![capture d ecran_620](https://user-images.githubusercontent.com/13449658/48543165-3debdd00-e8c1-11e8-87bb-58a35169192b.png)

marionf commented:
@TristanLDD there are many things on this page, tell me when you want to look at it

sarjon committed:
fix url to order view page

sarjon committed:
fix escaped string encdoing

sarjon commented:
@marionf thanks for feedback, issues were fixed. :)

marionf commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Customer/Blocks/View/orders.html.twig:
```diff
@@ -0,0 +1,139 @@
+                  <td>{{ order.orderProductsCount }}</td>
+                  <td>{{ order.totalPaid }}</td>
+                  <td class="text-right">
+                    <a href="{{ getAdminLink('AdminOrders', true, {'id_order': order.orderId, 'viewcustomer': 1}) }}"
```
@sarjon can you change it also here ?
```suggestion
                    <a href="{{ getAdminLink('AdminOrders', true, {'id_order': order.orderId, 'vieworder': 1}) }}"
```

sarjon replied:
done. :+1:

marionf commented:
Thank you @sarjon

marionf commented:
1) Voucher block :
- disabled vouchers are not displayed, they should be
- logo for enabled vouchers should be green like on the product page

![capture d ecran_633](https://user-images.githubusercontent.com/13449658/48616951-604f2a80-e995-11e8-9330-600b2538b7b2.png)

- logo for disabled vouchers should be red like on the product page

![capture d ecran_634](https://user-images.githubusercontent.com/13449658/48616962-680ecf00-e995-11e8-8d6a-fc0e2a284fe6.png)

2) Private note
- Delete the floppy disk of the save button, you can take this one:
https://cdn.rawgit.com/PrestaShop/prestashop-ui-kit/master/index.html

![save](https://user-images.githubusercontent.com/13449658/48617145-e4091700-e995-11e8-8f4d-191e606ff2de.png)

- Put the save button on the right 
- Put +10px between the textarea and the save button

3) Vouchers & Addresses
- Put edit and delete like on the product page

![capture d ecran_636](https://user-images.githubusercontent.com/13449658/48617501-e7e96900-e996-11e8-827f-596138168ac5.png)

4) Orders:
- Keep only the magnifying glass with a link to the order:
![capture d ecran 2018-11-16 a 12 14 26](https://user-images.githubusercontent.com/27484094/48618484-d0f84600-e999-11e8-9a2c-1abc46298add.png)

- If I click anywhere on the line, I am redirected to the order

5) Carts:
- Put only the magnifying glass with a link to the cart:
- Remove the link on the date
- If I click anywhere on the line, I am redirected to the cart

6) For orders, carts, vouchers, addresses
- Add and move the label "Actions" on the right

![capture d ecran_637](https://user-images.githubusercontent.com/13449658/48617823-c5a41b00-e997-11e8-994a-15262f84bf63.png)

sarjon committed:
fix private note ui

sarjon committed:
update lists UI

sarjon committed:
make orders, carts, addresses and discounts list rows clickable

sarjon committed:
rename customer viewing command

sarjon commented:
@marionf UI should be okay now. :)

sarjon committed:
fix key duplication in yaml

mickaelandrieu commented:
>  It ensures that, if someone modifies the class and introduces a syntax error (missing ;)

I think this is already checked by `php -l`, isn't it?

@sarjon would you mind to dump assets again please? :+1:

marionf commented:
Thank you @sarjon 

Some issues: 
- When I edit an address, the page is reloaded 
- When I delete an address or a voucher, I don't have any confirmation message
I should have the "Successful deletion." message
- Disabled vouchers are not displayed

sarjon commented:
> Disabled vouchers are not displayed

regarding disabled vouchers, im not sure we can easily achieve that without modifying core code. :/ 

@mickaelandrieu wdyt? as i understand, if vouchers are disabled, then `CartRule::isFeatureActive()` returns false, even though there are vouchers. Im afraid this may introduce regressions.

marionf commented:
Hum, sorry I just checked on legacy controller and disabled vouchers were not displayed, my bad
So, no need to change that :)

sarjon commented:
no problem.

> When I delete an address, I don't have any confirmation message

since deleting (also editing) is still handled by legacy (AdminAddressController), cannot be easily passed to new pages. It will be fixed once Addresses are migrated. Are you okay with that?

> When I edit an address, the page is reloaded

Link is now fixed.

sarjon committed:
fix link to customer address edit page

marionf commented:
It's ok for me :)
Thank you @sarjon

PierreRambaud commented:
Let's go @sarjon it will be merge after your rebase :D

sarjon committed:
build assets

sarjon commented:
rebased. :)

sarjon commented:
@PierreRambaud can you merge? ^^

Quetzacoalt91 reviewed (approved):
Let's merge before the next conflicts.

Quetzacoalt91 commented:
Thank you @sarjon

gomonkey commented:
Where can I find the view now? I need to add more info, I try to search it but without success.
Thank you
